backend/app/providers/telegram.go

```diff
@@ -0,0 +1,65 @@
+package providers
+
+// This provider provides common logic, reused by Telegram auth and notifications verification
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+
+	"github.com/umputun/remark42/backend/app/notify"
+)
+
+type tgRequester interface {
+	Request(ctx context.Context, method string, b []byte, data interface{}) error
+}
+
+// TGUpdatesReceiver used to dispatch telegram updates to multiple receivers
+type TGUpdatesReceiver interface {
+	ProcessUpdate(ctx context.Context, textUpdate string) error
+}
+
+// DispatchTelegramUpdates dispatches telegram updates to provided list of receivers
+// Blocks caller
+func DispatchTelegramUpdates(ctx context.Context, requester tgRequester, receivers []TGUpdatesReceiver, period time.Duration) {
+	var updateOffset int
+
+	processUpdatedTicker := time.NewTicker(period)
+	for {
+		select {
+		case <-ctx.Done():
+			processUpdatedTicker.Stop()
+			return
+		case <-processUpdatedTicker.C:
+			url := `getUpdates?allowed_updates=["message"]`
+			if updateOffset != 0 {
+				url += fmt.Sprintf("&offset=%d", updateOffset)
+			}
+
+			var result notify.TelegramUpdate
+
+			err := requester.Request(ctx, url, nil, &result)
+			if err != nil {
+				log.Printf("[ERROR] failed to fetch updates: %v", err)
+			}
+
+			for _, u := range result.Result {
```

Comment:
from the first sight I don't understand what this loop is doing. I see it iterates over some results and update offset. The part what is not clear - why some results here with UpdateID below the updateOffset you explicitly set above? And the second question - you seems to update offset in some cases but not altering result? Not sure what it means as I don't really understand all this loop in the first place. Maybe some little comment may clarify the mystery to me ;)